File: Go/Go_offer/offerII029.go
// Go
// leetcode offerII 029
// https://leetcode.cn/problems/4ueAj6/

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 * }
 */

func putAfter(aNode, bNode *Node) {
	bNode.Next = aNode.Next
	aNode.Next = bNode
}

func insert(he *Node, x int) *Node {
	tmp := &Node{Val : x, }
	// from 0 to 1 Node
    if he == nil {
    	tmp.Next = tmp
    	return tmp
    }

    // get min and max from the linklist
    res, min, max, cur := he, 2_000_000, -2_000_000, 0
    for {
    	cur = he.Val
    	if cur < min {
    		min = cur
    	}
    	if cur > max {
    		max = cur
    	}
    	he = he.Next
    	if he == res {
    		break
    	}
    }

    if min == max {
    	putAfter(res, tmp)
    	return res
    } else if x <= min || x >= max {
    	for !(he.Val == max && he.Next.Val == min) {
    		he = he.Next
    	}
    	putAfter(he, tmp)
    	return res
    } else {
    	// fmt.Println("here")
    	for !(he.Val <= x && he.Next.Val > x) {
    		he = he.Next
    	}
    	putAfter(he, tmp)
    	return res
    }
    return res // never
}

File: cpp/cpp_offer/offerII029.cc
// C++
// leetcode offerII 029
// https://leetcode.cn/problems/4ueAj6/

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;

    Node() {}

    Node(int _val) {
        val = _val;
        next = NULL;
    }

    Node(int _val, Node* _next) {
        val = _val;
        next = _next;
    }
};
*/

class Solution {
public:
	// put b after a in a linklist
	void ins(Node* a, Node* b) {
		Node* tmp = a->next;
		a->next = b;
		b->next = tmp;
	}

    Node* insert(Node* he, int x) {
    	Node *tmp = new Node(x), *res = he;
    	// n = 0
        if(he == nullptr) {
        	tmp->next = tmp;
        	return tmp;
        }

        int min_val = he->val, max_val = he->val;
        do {
        	min_val = min(min_val, he->val);
        	max_val = max(max_val, he->val);
        	he = he->next;
        } while (he != res);

        int v = 0;
        if(min_val == max_val) {
        	// one value for all nodes
        	ins(res, tmp);
        	return res;
        } else if (x <= min_val || x >= max_val) {
        	// max_val -> x -> min_val
        	while(!(he->val == max_val && he->next->val == min_val)) {
        		he = he->next;
        	}
        	ins(he, tmp);
        	return res;
        } else {
        	while(!(he->val <= x && he->next->val > x)) {
        		he = he->next;
        	}
        	ins(he, tmp);
        	return res;
        }
        return res;
    }
};